docs/gangplank/usage.md

```diff
@@ -1,3 +1,239 @@
+
+When `builds.json` and `meta.json` are found, Gangplank will always transmit these files to the workers. `kola` test results, logs and all `meta.json` files are always returned to the origin Gangplank.
+
+### Build Short-hands
+
+To support disitrbuted builds, Gangplank has two special build shorthands of "base" and "finalize":
+
+```
+stages:
+ - id: base
+   build_artifacts:
+   - base
+   - finalize
+```
+
+The `base` short-hand corresponds to `cosa build --delay-meta-merge`, while `finalize` corresponds to `cosa meta --finalize`. By default, `cosa build` uses the `delay-meta-merge` since Gangplank is performing a distributed build. In general `finalize` should be the last stage.
+
+### JobSpec
+
+The JobSpec (or Job Specification) is simply YAML that instruct Gangplank on the steps and dependencies for starting a build.
+
+To get started with a JobSpec, you can generate one using Gangplank.:
+```
+$ bin/gangplank generate -A base
+INFO[0000] Gangplank: COSA OpenShift job runner, 2021-03-02.9dce8136~dirty
+# Generated by Gangplank CLI
+# 2021-03-02T17:25:42-07:00
+job:
+  strict: true
+recipe:
+  git_ref: testing-devel
+  git_url: https://github.com/coreos/fedora-coreos-config
+stages:
+- id: Generated Base Stage
+  build_artifacts: [base]
+- id: Generated Finalize Stage
+  build_artifacts: [finalize]
+delay_meta_merge: true
+```
+
+The JobSpec defines discrete, units of work as a "stage". Each stage supports few options:
+
+- id: the name of the stage; must be unique
+- command: a list of commands to execute
+- concurrent_execution: bool to indicate if the `commands` can be executed concurrently
+- build_artifacts: known artifacts to build
+- direct_execution: do not apply templating
+- prep_commands: a list of commands to run before command
+- post_commands: a list of commands to run last (such as test or cleanup)
+- post_always: a bool that indicates whether the `post_commands` should _always be executed regardless of the success of the `commands` stage.
+- require_artifacts: the type of artifact that's require for work (i.e. `qemu` or `aws`). Stages will not start until the artifact appears.
+
+To illustrate this, consider:
+```
+- id: base Stage
+  build_artifacts:
+  - base
+  post_commands:
+  - cosa kola run --basic-qemu-scenarios
+- id: clouds
+  concurrent_execution: true
+  build_artifacts:
+  - aws
+  - gcp
+  required_artifacts:
+  - qemu
+- id: finalize
+  build_artifacts:
+  - finalize
+```
+
+In this example, Gangplank will:
+
+1. Build and test base artifacts in the first pod
+1. In a new pod, concurrently build AWS and GCP but only after the QEMU artifact is found.
+1. Combine meta.json together
+
+### Meta Data and JobSpec Templating
+
+Gangplank was initially started after belately realizing that the Jenkins Pipelines are, in fact, complicated templating engines. That is, a considerable amount of time, energy and development was put into translating data from YAML/JSON into execution rules.
```

Comment:
Right, Prow also had this problem a lot but as of this year the team invested in a [multi-stage step workflow](https://docs.ci.openshift.org/docs/architecture/step-registry/) that's used by a lot of jobs.

I think probably then, Gangplank-in-Prow would be a step that basically just calls gangplank/cosa?

Reply:
Gangplank-in-Prow would be exactly calls to Gangplank from Prow to run COSA.

For Prow, I see multi-stage step worklfows and Gangplank working together -- Prow can start the Gangplank workers, but Gangplank may create several pods for doing the actual work.